oauth2_client/models.py

```diff
@@ -1,54 +1,259 @@
 """
 Ouath2 client models.
 """
-from django.contrib.postgres.fields import JSONField
 from django.db import models
+from django.forms import model_to_dict
 from django.utils import timezone
+from django.conf import settings
+from datetime import timedelta
+
+from django.utils.translation import ugettext_lazy as _
+from django.contrib.postgres.fields import JSONField
+
+from labsterutils.obj import get_value_or_default
+from oauth2_client.exceptions import TokenError
 
 
 class Application(models.Model):
     """
-    Client credentials for the service provider.
+    Client Application model
     """
-    name = models.CharField(max_length=50, help_text="Example: licence")
-    # base definitions copied from oauth2_provider.Application
-    client_id = models.CharField(max_length=100)
-    client_secret = models.CharField(max_length=100)
+    CLIENT_CONFIDENTIAL = "confidential"
+    CLIENT_PUBLIC = "public"
+    CLIENT_TYPES = (
+        (CLIENT_CONFIDENTIAL, _("Confidential")),
+        (CLIENT_PUBLIC, _("Public")),
+    )
+
+    GRANT_AUTHORIZATION_CODE = "authorization-code"
+    GRANT_IMPLICIT = "implicit"
+    GRANT_PASSWORD = "password"
+    GRANT_CLIENT_CREDENTIALS = "client-credentials"
+    GRANT_JWT_BEARER = "jwt-bearer"
+    GRANT_TYPES = (
+        (GRANT_AUTHORIZATION_CODE, _("Authorization code")),
+        (GRANT_IMPLICIT, _("Implicit")),
+        (GRANT_PASSWORD, _("Resource owner password-based")),
+        (GRANT_CLIENT_CREDENTIALS, _("Client credentials")),
+        (GRANT_JWT_BEARER, _("JWT bearer")),
+    )
+
+    id = models.BigAutoField(primary_key=True)
+    provider = models.CharField(
+        verbose_name=_('provider'),
+        max_length=30
+    )
+    name = models.CharField(
+        max_length=255,
+        blank=False,
+        unique=True
+    )
+    client_id = models.CharField(
+        max_length=100,
+    )
+    client_secret = models.CharField(
+        max_length=255,
+        blank=True,
+    )
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        related_name="%(app_label)s_%(class)s",
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE
+    )
+    client_type = models.CharField(
+        max_length=32,
+        choices=CLIENT_TYPES
+    )
+    authorization_grant_type = models.CharField(
+        max_length=32,
+        choices=GRANT_TYPES
+    )
+
+    created = models.DateTimeField(auto_now_add=True)
+    updated = models.DateTimeField(auto_now=True)
```

Comment:
we stick to `created_at`, `updated_at` naming in our code base

Reply:
changing